src/modules/job-manager/plugins/perilog.c

```diff
@@ -0,0 +1,513 @@
+     *  next fulfillment will be the timeout set by the original call to
+     *  flux_future_then() below. If the process is destroyed before the
+     *  timer expires, then the future will also be destroyed so the
+     *  timeout above won't occur.
+     */
+    flux_future_reset (f);
+}
+
+
+static int prolog_kill (struct perilog_proc *proc)
+{
+    flux_t *h = flux_jobtap_get_flux (proc->p);
+
+    if (!(proc->kill_f = flux_subprocess_kill (proc->sp, SIGTERM)))
+        return -1;
+
+    if (flux_future_then (proc->kill_f,
+                          perilog_config.prolog_kill_timeout,
+                          prolog_kill_cb, proc) < 0) {
+        flux_log_error (h, "prolog_kill: flux_future_then");
+        flux_future_destroy (proc->kill_f);
+        proc->kill_f = NULL;
+        return -1;
+    }
+
+    return 0;
+}
+
+static int exception_cb (flux_plugin_t *p,
+                         const char *topic,
+                         flux_plugin_arg_t *args,
+                         void *arg)
+{
+    /*  On exception, kill any prolog running for this job:
+     *   Follow up with SIGKILL after 10s.
+     */
+    struct perilog_proc *proc;
+    if ((proc = flux_jobtap_job_aux_get (p,
+                                        FLUX_JOBTAP_CURRENT_JOB,
+                                        "perilog_proc"))
+        && flux_subprocess_state (proc->sp) == FLUX_SUBPROCESS_RUNNING)
+        return prolog_kill (proc);
+    return 0;
+}
+
+static flux_cmd_t *cmd_from_json (json_t *o)
+{
+    size_t index;
+    json_t *value;
+    flux_cmd_t *cmd;
+
+    if (!json_is_array (o))
+        return NULL;
+
+    if (!(cmd = flux_cmd_create (0, NULL, environ)))
+        return NULL;
+
+    json_array_foreach (o, index, value) {
+        const char *arg = json_string_value (value);
+        if (!value
+            || flux_cmd_argv_append (cmd, arg) < 0)
+            goto fail;
+    }
+    return cmd;
+fail:
+    flux_cmd_destroy (cmd);
+    return NULL;
+}
+
+/*  Parse [job-manager.prolog] and [job-manager.epilog] config
+ */
+static int conf_init (flux_t *h, struct perilog_conf *conf)
+{
+    flux_conf_error_t error;
+    json_t *prolog = NULL;
+    json_t *epilog = NULL;
+
+    memset (conf, 0, sizeof (*conf));
+    conf->prolog_kill_timeout = 5.;
+    if (!(conf->processes = job_hash_create ()))
+        return -1;
+    zhashx_set_destructor (conf->processes,
+                           perilog_proc_destructor);
+
+    if (flux_conf_unpack (flux_get_conf (h),
+                          &error,
+                          "{s?:{s?:{s?o s?F !} s?:{s?o !}}}",
+                          "job-manager",
+                          "prolog",
+                            "command", &prolog,
+                            "kill-timeout", &conf->prolog_kill_timeout,
+                          "epilog",
+                            "command", &epilog) < 0) {
+        flux_log_error (h, "prolog/epilog configuration error: %s",
```

Comment:
`flux_log (h, LOG_ERR, ...` might be more appropriate here (error string appended by `flux_log_error()` is likely redundant with error.errbuf)